Add enable and disable commands

bronson opened the pull request:
Addresses https://github.com/atom/apm/issues/358 and https://github.com/atom/atom/issues/5987.

There's some code duplication but that seems to be apm's style: commands and tests are mostly self-sufficient.  I wanted to make the new commands resemble the existing ones.

Fishrock123 commented:
I'm not sure if it is within the domain of this, but would there be a way to use this to make sure disabled packages are not re-enabled on upgrade?

bronson commented:
I haven't seen that happen...  I've seen a corruption of ~/.atom/config.cson cause it to be replaced with the defaults, re-enabling all the packages, but that was a sharing-config-on-dropbox issue.

Do you have an issue # for it?

mnquintana commented:
@Fishrock123 That's an unrelated settings-view issue - I can't seem to find the issue # right now though.

mnquintana commented:
This looks pretty good! I did notice though that it's currently possible to pass in a non-existent package name to both `enable` and `disable` (`disable` will add it to the `core.disabledPackages` list) – would you mind logging an error for any non-existent package names that are passed to these commands? (And adding specs for that?)

mnquintana commented on src/disable.coffee:
```diff
@@ -0,0 +1,56 @@
+      callback "Failed to save `#{configFilePath}`: #{error.message}"
+      return
+
+    process.stdout.write "Disabled: #{packageNames.join(', ')} "
```
Pretty minor, but I think we can just use `console.log` here (I know it's `process.stdout.write` elsewhere in apm), although they're so similar that it doesn't really matter

mnquintana replied:
What do you think about having a newline for each disabled package? I think it would make it a bit more readable if you're disabling a bunch of commands at once, and easier to provide more granular feedback to the user (e.g. if one of the packages was non-existent or something)

bronson replied:
```
~/apm (master)⚡$ grep -r stdout.write src | wc -l
      36
~/apm (master)⚡$ grep -r console.log src | wc -l
      41
```

console.log wins by a hair!  Agreed, converting to console.log.

bronson replied:
Now it outputs like this:

~/apm $ bin/apm enable fee fi fo fum aiee
Enabled:
  fee
  fi
  fo
  fum
  aiee 
✓

Better?

mnquintana commented on src/disable.coffee:
```diff
@@ -0,0 +1,56 @@
+      return
+
+    keyPath = '*.core.disabledPackages'
+    disabledPackages = _.valueForKeyPath(settings, keyPath) or []
```
:art: We usually use the existential operator `?` for this

bronson replied:
Unless I'm missing something, that won't work.  If I change that line to:

```coffeescript
disabledPackages = _.valueForKeyPath(settings, keyPath)? or []
```
Then the return value ends up either true or false -- the disabledPackages gets thrown away.  Right?

mnquintana replied:
It's been awhile now :sweat_smile: but I think I meant this:

```coffee
disabledPackages = _.valueForKeyPath(settings, keyPath) ? []
```

It's really just a :art: thing since in this case the existential operator works exactly the same as the conditional assignment you have here.

bronson replied:
Ah, gotcha.  Yep, that compiles down to

```js
disabledPackages = (ref = _.valueForKeyPath(settings, keyPath)) != null ? ref : [];
```

Which looks and works great.  Change made.

bronson commented:
@mnquintana hope I've addressed all your concerns.

It might seem a little odd that the Disable command calls into the List command to retrieve the list of installed packages, but I couldn't find a better way of doing it that didn't duplicate a ton of the List command's code.  One way would be to extract the listing functionality out of the List command and into a shared module but that's a pretty serious change.  This seems the least evil of the alternatives.  :)

New thoughts?

mnquintana commented:
@bronson Sorry for the super long delay in re-reviewing this - taking a look now

bronson commented:
@mnquintana how's it look?  :)

mnquintana commented on src/disable.coffee:
```diff
@@ -0,0 +1,83 @@
+
+    configFilePath = CSON.resolve(path.join(config.getAtomDirectory(), 'config'))
+    unless configFilePath
+      callback("Could not find config.cson. Run Atom first?")
```
It might be a good idea to add a spec for this case (when there is no `config.cson`)

mnquintana replied:
Just kidding, there totally is one already :+1:

lee-dohm commented:
Looks good to me! :ship:

mnquintana commented:
Sorry about the delay here @bronson, but thanks so much for adding these! apm just got significantly more powerful :rocket:

bronson commented:
Yay, thanks!

modulexcite commented:
Has this been released into production yet?

mnquintana commented:
@modulexcite It has! It should work with the latest production version of Atom.

modulexcite commented:
Thanks for your quick response.Would be kind enough to tell me what the command is to list all installed but disabled/enabled packages from apm.

bronson commented:
I don't think there's a way to do it explicitly.  `grep -v (disabled)` (or some ()-escaped variant) should do it now.

It probably wouldn't be too hard to add an `apm list --disabled` or `apm list --enabled` options.

modulexcite commented:
Sorry I forgot to mention that I'm on Windows 7 OS and can't do grep.

bronson commented:
Well then...   I vote for submitting a pull request.  :)

bronson commented:
Or, heck, I'll submit one myself: #623

`apm list --disabled` is actually pretty handy.

modulexcite commented:
That would be awesome, thanks.
